error.go

```diff
@@ -72,22 +65,242 @@ func collectMessages(data interface{}, messages []string) []string {
+	Errors []*Error `json:"errors,omitempty"`
 }
 
-// Error provides a pretty printed string of an error object based on present data.
+// Error provides a pretty printed string of an APIError object based on present data.
 func (e *APIError) Error() string {
 	if e.Message != "" {
 		if e.Code != "" {
-			return e.Code + " " + e.Message.(string)
+			return e.Code + " " + e.Message
 		}
-		return e.Message.(string)
+		return e.Message
 	}
 	if e.Code != "" {
 		return e.Code
 	}
-	return fmt.Sprintf("%d %s", e.StatusCode, e.Status)
+	return fmt.Sprintf("%d %s", e.StatusCode, e.Code)
+}
+
+type BadRequestError struct {
+	APIError
+}
+
+type ConnectionError struct {
+	APIError
+}
+
+type GatewayTimeoutError struct {
+	APIError
+}
+
+type InternalServerError struct {
+	APIError
+}
+
+type InvalidRequestError struct {
+	APIError
+}
+
+type MethodNotAllowedError struct {
+	APIError
+}
+
+type NotFoundError struct {
+	APIError
+}
+
+type PaymentError struct {
+	APIError
+}
+
+type ProxyError struct {
+	APIError
+}
+
+type RateLimitError struct {
+	APIError
+}
+
+type RedirectError struct {
+	APIError
+}
+
+type RetryError struct {
+	APIError
+}
+
+type ServiceUnavailableError struct {
+	APIError
+}
+
+type SSLError struct {
+	APIError
+}
+
+type TimeoutError struct {
+	APIError
+}
+
+type UnauthorizedError struct {
+	APIError
 }
 
-var EndOfPaginationError = errors.New("there are no more pages to retrieve")
+type UnknownHttpError struct {
+	APIError
+}
+
+// FromErrorResponse returns an APIError-based object based on the HTTP response.
+// Do not pass a non-error response to this function.
+func FromErrorResponse(response *http.Response) error {
+	// build the base APIError object from the response
+	apiError := &APIError{
+		StatusCode: response.StatusCode,
+	}
+
+	// deserialize the response body into a temporary object
+	buf, _ := ioutil.ReadAll(response.Body)
+	aux := &struct {
+		Error *Error `json:"error,omitempty"`
+	}{}
+
+	if json.Unmarshal(buf, &aux) == nil {
+		// extract the details from the temporary object (top-level Error class) and store them in the APIError object
+		apiError.Message = aux.Error.Message.(string)
+		apiError.Code = aux.Error.Code
+		apiError.Errors = aux.Error.Errors
+	} else {
+		// could not extract error details from the API response (or API did not return data, i.e. 1xx, 3xx or 5xx)
+		if response.Status == "" {
+			response.Status = ApiDidNotReturnErrorDetails
+		}
+		apiError.Message = response.Status
+		apiError.Code = ApiErrorDetailsParsingError
+		apiError.Errors = []*Error{}
+	}
+
+	// return the appropriate error type based on the status code
+	switch response.StatusCode {
+	case 0:
+		return &ConnectionError{APIError: *apiError}
+	case 100, 101, 102, 103:
+		return &RetryError{APIError: *apiError}
+	case 300, 301, 302, 303, 304, 305, 306, 307, 308:
+		return &RedirectError{APIError: *apiError}
+	case 400:
+		return &BadRequestError{APIError: *apiError}
+	case 401, 403:
```

Comment:
issue: 401 is unauthorized, 403 is forbidden and the distinction is important. Can we break these two up like we've done elsewhere?

Reply:
Resolved